File: LeetCode/Medium/0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.go
func characterReplacement(s string, k int) int {
	//AABABBA
	//map - A{1}
	//countMaxfreq - 1
	//windowLength - 1
	countMaxfreq, maxLength, left := 0, 0, 0
	windowMap := make(map[byte]int)

	for right := 0; right < len(s); right++ {
		windowLength := right - left + 1
		windowMap[s[right]]++
		countMaxfreq = max(countMaxfreq, windowMap[s[right]])
		if (windowLength - countMaxfreq) > k {
			windowMap[s[left]]--
			left++
		} else {
			maxLength = max(maxLength, windowLength)
		}
	}
	return maxLength
}

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

File: LeetCode/Medium/0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char, int> ump;
        int left = 0, maxCount = 0, winLen = 0, maxLen = 0;
        for (int right = 0; right < s.size(); right++) {
            ump[s[right]]++;
            maxCount = max(maxCount, ump[s[right]]);
            winLen = right - left + 1;
            if (winLen - maxCount <= k) {
                maxLen++;
            } else {
                ump[s[left]]--;
                left++;
            }
        }
        return maxLen;
    }
};